### ISSUE TITLE
NCBI Entrez protein IPG XML issue, XSD missing Protein and ProteinList

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#1109]
At the time of writing, our test ``test_Entrez_online.py`` fails as follows:

```
$ python3 test_Entrez_online.py
...
======================================================================
ERROR: test_fetch_xml_schemas (test_Entrez_online.EntrezOnlineCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/pc40583/repositories/biopython/Tests/test_Entrez_online.py", line 207, in test_fetch_xml_schemas
    records = list(Entrez.parse(handle))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/Bio/Entrez/Parser.py", line 286, in parse
    self.parser.Parse(text, False)
  File "/Users/sysadmin/build/v3.6.0/Modules/pyexpat.c", line 414, in StartElement
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/Bio/Entrez/Parser.py", line 368, in startElementHandler
    raise ValidationError(name)
Bio.Entrez.Parser.ValidationError: Failed to find tag 'ProteinList' in the DTD. To skip all tags that are not represented in the DTD, please call Bio.Entrez.read or Bio.Entrez.parse with validate=False.

----------------------------------------------------------------------
```

We can verify this outside of Biopython, this is https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&id=783730874&rettype=ipg&retmode=xml downloaded today,

```xml
$ more protein_ipg.xml 
<?xml version="1.0" ?>
<IPGReportSet xmlns:xsi="https://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://www.ncbi.nlm.nih.gov/data_specs/schema/other/seq_report/IPGReportSet.xsd">

<IPGReport  ipg="79092155" product_acc="KJV04014.1">
<Product  accver="WP_045840896.1" name="methylmalonate-semialdehyde dehydrogenase (CoA acylating)" taxid="1356405" slen="513" org="Rhodococcus sp. PML026" kingdom_taxid="2" kingdom="Bacteria"/>
<ProteinList>
<Protein  accver="KJV04014.1" source="INSDC" name="methylmalonic acid semialdehyde dehydrogenase mmsa" taxid="1356405" org="Rhodococcus sp. PML026" kingdom_taxid="2" kingdom="Bacteria" priority="0">
<CDSList>
<CDS  accver="JZIS01000004.1" start="264437" stop="265978" strand="-" taxid="1356405" org="Rhodococcus sp. PML026" kingdom_taxid="2" kingdom="Bacteria" strain="PML026"/>
<CDS  accver="KQ031368.1" start="264437" stop="265978" strand="-" taxid="1356405" org="Rhodococcus sp. PML026" kingdom_taxid="2" kingdom="Bacteria" strain="PML026"/></CDSList></Protein>
<Protein  accver="WP_045840896.1" source="RefSeq" name="methylmalonate-semialdehyde dehydrogenase (CoA acylating)" taxid="1356405" org="Rhodococcus sp. PML026" kingdom_taxid="2" kingdom="Bacteria" priority="1">
<CDSList>
<CDS  accver="NZ_KQ031368.1" start="264437" stop="265978" strand="-" taxid="1356405" org="Rhodococcus sp. PML026" kingdom_taxid="2" kingdom="Bacteria" strain="PML026"/></CDSList></Protein></ProteinList>
<Statistics  prot_count="2" nuc_count="3"/></IPGReport></IPGReportSet>

```

This references XSD file https://www.ncbi.nlm.nih.gov/data_specs/schema/other/seq_report/IPGReportSet.xsd again downloaded today:

```xml
$ more IPGReportSet.xsd 
<U+FEFF><?xml version="1.0" ?>
<xs:schema
  xmlns:xs="http://www.w3.org/2001/XMLSchema"
  xmlns:ncbi="http://www.ncbi.nlm.nih.gov"
  elementFormDefault="unqualified"
  attributeFormDefault="unqualified"
>
    <xs:element name="CDS">
        <xs:complexType>
            <xs:attribute name="accver" type="xs:string" use="required"/>
            <xs:attribute name="gi" type="xs:integer" use="optional"/>
            <xs:attribute name="kingdom" type="xs:string" use="required"/>
            <xs:attribute name="kingdom_taxid" type="xs:integer" use="required"/>
            <xs:attribute name="org" type="xs:string" use="required"/>
            <xs:attribute name="start" type="xs:integer" use="required"/>
            <xs:attribute name="stop" type="xs:integer" use="required"/>
            <xs:attribute name="strand" type="xs:string" use="required"/>
            <xs:attribute name="strain" type="xs:string" use="required"/>
            <xs:attribute name="taxid" type="xs:integer" use="required"/>
        </xs:complexType>
    </xs:element>

    <xs:element name="CDSList">
        <xs:complexType>
            <xs:sequence maxOccurs="unbounded">
                <xs:element ref="CDS"/>
            </xs:sequence>
        </xs:complexType>
    </xs:element>

    <xs:element name="IPGReport">
        <xs:complexType>
            <xs:sequence>
                <xs:element ref="Product"/>
                <xs:element ref="RedundantGiList"/>
                <xs:element ref="Statistics"/>
            </xs:sequence>
            <xs:attribute name="accver" type="xs:string" use="required"/>
            <xs:attribute name="gi" type="xs:integer" use="optional"/>
            <xs:attribute name="kingdom" type="xs:string" use="required"/>
            <xs:attribute name="kingdom_taxid" type="xs:integer" use="required"/>
            <xs:attribute name="org" type="xs:string" use="required"/>
            <xs:attribute name="taxid" type="xs:integer" use="required"/>
        </xs:complexType>
    </xs:element>

    <xs:element name="IPGReportSet">
        <xs:complexType>
            <xs:sequence>
                <xs:element ref="IPGReport"/>
            </xs:sequence>
        </xs:complexType>
    </xs:element>

    <xs:element name="Product">
        <xs:complexType>
            <xs:simpleContent>
                <xs:extension base="xs:string">
                    <xs:attribute name="accver" type="xs:string" use="required"/>
                    <xs:attribute name="gi" type="xs:integer" use="optional"/>
                    <xs:attribute name="slen" type="xs:integer" use="required"/>
                </xs:extension>
            </xs:simpleContent>
        </xs:complexType>
    </xs:element>

    <xs:element name="RedundantGi">
        <xs:complexType>
            <xs:sequence minOccurs="0">
                <xs:element ref="CDSList"/>
            </xs:sequence>
            <xs:attribute name="accver" type="xs:string" use="required"/>
            <xs:attribute name="gi" type="xs:integer" use="optional"/>
            <xs:attribute name="name" type="xs:string" use="required"/>
            <xs:attribute name="kingdom" type="xs:string" use="required"/>
            <xs:attribute name="kingdom_taxid" type="xs:integer" use="required"/>
            <xs:attribute name="org" type="xs:string" use="required"/>
            <xs:attribute name="priority" type="xs:string" use="required"/>
            <xs:attribute name="source" type="xs:string" use="required"/>
            <xs:attribute name="taxid" type="xs:integer" use="required"/>
        </xs:complexType>
    </xs:element>

    <xs:element name="RedundantGiList">
        <xs:complexType>
            <xs:sequence maxOccurs="unbounded">
                <xs:element ref="RedundantGi"/>
            </xs:sequence>
        </xs:complexType>
    </xs:element>

    <xs:element name="Statistics">
        <xs:complexType>
            <xs:attribute name="nuc_count" type="xs:integer" use="required"/>
            <xs:attribute name="prot_count" type="xs:integer" use="required"/>
        </xs:complexType>
    </xs:element>

</xs:schema>

```

As reported by our Entrez parser, the XML element ``ProteinList`` is not defined in the XSD file. This appears therefore to be a (new/recent) NCBI problem. So too is ``Protein`` as shown here:

```
$ ./entrez_dtd_bug.py 
XML file references the XSD here:
<IPGReportSet xmlns:xsi="https://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://www.ncbi.nlm.nih.gov/data_specs/schema/other/seq_report/IPGReportSet.xsd">

The XSD defines the following elements:
 - CDS
 - CDSList
 - IPGReport
 - IPGReportSet
 - Product
 - RedundantGi
 - RedundantGiList
 - Statistics
The XML uses the following elements
 - CDS
 - CDSList
 - IPGReport
 - IPGReportSet
 - Product
 - Protein (UNDEFINED)
 - ProteinList (UNDEFINED)
 - Statistics
```

Using this self-contained example ``entrez_dtd_bug.py``,

```python
#!/usr/bin/env python3
import os
from urllib.request import urlretrieve
import xml.etree.ElementTree as ET

xml_url = "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&id=783730874&rettype=ipg&retmode=xml"
xml_file = "protein_ipg.xml"

if not os.path.isfile(xml_file):
    urlretrieve(xml_url, xml_file)

xsd_url = "https://www.ncbi.nlm.nih.gov/data_specs/schema/other/seq_report/IPGReportSet.xsd"
xsd_file = "IPGReportSet.xsd"

if not os.path.isfile(xsd_file):
    urlretrieve(xsd_url, xsd_file)


with open(xml_file) as h:
    for line in h:
        if xsd_url in line:
            print("XML file references the XSD here:")
            print(line)

def defined_tags(xsd_file):
    tags = set()
    with open(xsd_file) as h:
        for element in ET.fromstring(h.read()).iter():
            if "element" in element.tag and "name" in element.attrib:
                tags.add(element.attrib["name"])
    return sorted(tags)

def used_tags(xml_file):
    tags = set()
    with open(xml_file) as h:
        for element in ET.fromstring(h.read()).iter():
            tags.add(element.tag)
    return sorted(tags)

xsd_elements = defined_tags(xsd_file)
print("The XSD defines the following elements:")
for name in xsd_elements:
    print(" - %s" % name)

print("The XML uses the following elements")
xml_elements = used_tags(xml_file)
for name in xml_elements:
    if name in xsd_elements:
        print(" - %s" % name)
    else:
        print(" - %s (UNDEFINED)" % name)

```

I will log this with the NCBI, and CC @mdehoon

**Comment by peterjc** [comment]
The NCBI replied the XSD has now been updated...

**Comment by peterjc** [comment]
Good news, https://www.ncbi.nlm.nih.gov/data_specs/schema/other/seq_report/IPGReportSet.xsd does now include:

```xml
  <xs:element name="Protein">
    <xs:complexType>
      <xs:sequence minOccurs="0">
        <xs:element ref="CDSList"/>
      </xs:sequence>
      <xs:attribute name="accver" type="xs:string" use="required"/>
      <xs:attribute name="source" type="xs:string" use="required"/>
      <xs:attribute name="name" type="xs:string" use="required"/>
      <xs:attribute name="org" type="xs:string" use="required"/>
      <xs:attribute name="kingdom" type="xs:string" use="required"/>
      <xs:attribute name="kingdom_taxid" type="xs:integer" use="required"/>
      <xs:attribute name="taxid" type="xs:integer" use="optional"/>
      <xs:attribute name="priority" type="xs:string" use="required"/>
    </xs:complexType>
  </xs:element>

  <xs:element name="ProteinList">
    <xs:complexType>
      <xs:sequence maxOccurs="unbounded">
        <xs:element ref="Protein"/>
      </xs:sequence>
    </xs:complexType>
  </xs:element>
```

**Comment by peterjc** [comment]
Bad news, uncommenting the test gave this:

```
======================================================================
ERROR: test_fetch_xml_schemas (test_Entrez_online.EntrezOnlineCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/Bio/Entrez/Parser.py", line 379, in startElementHandler
    current.append(object)
AttributeError: 'str' object has no attribute 'append'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/pc40583/repositories/biopython/Tests/test_Entrez_online.py", line 207, in test_fetch_xml_schemas
    records = list(Entrez.parse(handle, validate=False))
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/Bio/Entrez/Parser.py", line 286, in parse
    self.parser.Parse(text, False)
  File "/Users/sysadmin/build/v3.6.0/Modules/pyexpat.c", line 414, in StartElement
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/Bio/Entrez/Parser.py", line 381, in startElementHandler
    current[name] = object
TypeError: 'str' object does not support item assignment

----------------------------------------------------------------------
```

I don't have time to look into this any further today.

**Comment by mdehoon** [comment]
Fixed by pull request #1642 .
